src/authentication/authentication-hapi-plugin.ts

```diff
@@ -38,23 +51,171 @@ class AuthenticationHapiPlugin extends HapiPlugin {
+      handler: {
+        async: this.createTeamTokenHandler,
+      },
+      config: {
+        bind: this,
+        auth: 'admin',
+        validate: {
+          params: {
+            teamId: Joi.number().required(),
+          },
+        },
+      },
+    });
     next();
   }
 
   public async getTeamHandler(request: Hapi.Request, reply: Hapi.IReply) {
-    const credentials = request.auth.credentials as AccessToken;
-    const user = await getUserByEmail(credentials[this.subEmailClaimKey], this.gitlab);
-    const teams = getUserTeams(user.id, this.gitlab);
-    return reply(teams);
+    try {
+      const credentials = request.auth.credentials as AccessToken;
+      const user = await this.gitlab.getUserByEmail(credentials[subEmailClaimKey]);
+      const teams = await this.gitlab.getUserTeams(user.id);
+      return reply(teams[0]); // NOTE: we only support a single team for now
+    } catch (error) {
+      this.logger.error(`Can't fetch user or team`, error);
+      return reply(Boom.wrap(error, 404));
+    }
+  }
+
+  public async getTeamTokenHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    try {
+      const teamId = parseInt(request.paramsArray[0], 10);
+      const teamToken: TeamToken[] = await teamTokenQuery(this.db, undefined, teamId);
+      if (!teamToken.length) {
+        throw new Error(`No token found for team ${teamId}`);
+      }
+      return reply(teamToken[0]);
+    } catch (error) {
+      return reply(Boom.notFound(error.message));
+    }
+  }
+
+  public async createTeamTokenHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    try {
+      const teamId = parseInt(request.paramsArray[0], 10);
+      const teamToken = await generateAndSaveTeamToken(teamId, this.db);
+      return reply(teamToken).code(201);
+    } catch (error) {
+      return reply(Boom.badRequest(error.message));
+    }
+  }
+
+  public async signupHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    try {
+      const credentials = request.auth.credentials as AccessToken;
+      const teamToken = credentials[teamTokenClaimKey]!;
+      const teamId = await getTeamIdWithToken(teamToken, this.db);
+      const team = await this.gitlab.fetchJson<Group>(`groups/${teamId}`);
+      const email = credentials[subEmailClaimKey];
+      const password = generatePassword();
+      const {id, idp} = parseSub(credentials.sub);
+      const user = await this.gitlab.createUser(
+        email,
+        password,
+        email.replace('@', '-'),
+        email,
+        id,
+        idp,
+      );
+      await this.gitlab.addUserToGroup(user.id, teamId);
+      return reply({
+        team,
+        password,
+      });
+    } catch (error) {
+      this.logger.error(`Problems on signup`, error);
+      return reply(Boom.wrap(error, 401));
+    }
   }
 
   public async validateUser(
     payload: AccessToken,
-    _request: any,
+    request: any,
     callback: (err: any, valid: boolean, credentials?: any) => void,
```

Comment:
This function is called by the `hapi-auth-jwt2` plugin and at least the documentation doesn't mention Promise support.